packages/account/src/BiconomySmartAccountV2.ts

```diff
@@ -499,11 +507,75 @@ export class BiconomySmartAccountV2 extends BaseSmartContractAccount {
+   * @param {PaymasterUserOperationDto} paymasterServiceData - The paymaster service data containing mode and additional information.
+   * @returns {Promise<Partial<UserOperationStruct>>} A promise that resolves to the modified user operation structure.
+   */
+  async setPaymasterFields(
+    userOp: Partial<UserOperationStruct>,
+    paymasterServiceData: PaymasterUserOperationDto,
+  ): Promise<Partial<UserOperationStruct>> {
+    if (this.paymaster !== undefined) {
+      try {
+        if (paymasterServiceData.mode === PaymasterMode.SPONSORED) {
+          const paymasterData = await (this.paymaster as IHybridPaymaster<PaymasterUserOperationDto>).getPaymasterAndData(userOp, {
+            mode: paymasterServiceData.mode,
+          });
+          userOp.paymasterAndData = paymasterData.paymasterAndData;
+          userOp.callGasLimit = paymasterData.callGasLimit;
+          userOp.verificationGasLimit = paymasterData.verificationGasLimit;
+          userOp.preVerificationGas = paymasterData.preVerificationGas;
+          return userOp;
+        } else if (paymasterServiceData.mode === PaymasterMode.ERC20) {
+          const feeQuotesResponse: FeeQuotesOrDataResponse = await (
+            this.paymaster as IHybridPaymaster<PaymasterUserOperationDto>
+          ).getPaymasterFeeQuotesOrData(userOp, {
+            mode: PaymasterMode.ERC20,
+            tokenList: paymasterServiceData.tokenList,
+            preferredToken: paymasterServiceData.preferredToken,
+          });
+          const finalUserOp = await this.buildTokenPaymasterUserOp(userOp, {
+            // @ts-expect-error There should always be a fee quote
+            feeQuote: feeQuotesResponse.feeQuotes[0]!,
+            spender: (feeQuotesResponse.tokenPaymasterAddress as Hex) || "",
+            maxApproval: true,
+          });
+          const newPaymasterServiceData = {
+            mode: PaymasterMode.ERC20,
+            // @ts-expect-error There should always be a fee quote tokenAddress
+            feeTokenAddress: feeQuotesResponse.feeQuotes[0].tokenAddress!, // TODO: check if this is correct
+            calculateGasLimits: true, // Always recommended and especially when using token paymaster
+          };
+          const paymasterAndDataWithLimits = await (this.paymaster as IHybridPaymaster<PaymasterUserOperationDto>).getPaymasterAndData(
+            finalUserOp,
+            newPaymasterServiceData,
+          );
+          finalUserOp.paymasterAndData = paymasterAndDataWithLimits.paymasterAndData;
+          finalUserOp.callGasLimit = paymasterAndDataWithLimits.callGasLimit;
+          finalUserOp.verificationGasLimit = paymasterAndDataWithLimits.verificationGasLimit;
+          finalUserOp.preVerificationGas = paymasterAndDataWithLimits.preVerificationGas;
+          return finalUserOp;
+        } else {
+          return userOp;
+        }
+      } catch (e: any) {
+        Logger.error("Error while fetching paymaster data", e);
+        return userOp;
```

Comment:
Should we return or just throw here? Difficult to understand when / why things should fail, but are we risking users submitting wrong values by returning here when it could error instead?

Reply:
I think both ways are ok, I chose to return user op to not add more return type to the function.

Reply:
I don't think you'll have to change the return type if you throw an error. Unless there's a good reason not to I would throw here, otherwise users won't know if something has failed and proceed assuming they have gotten the correct gas values. Wouldn't that be exactly the problem described in this thread? https://biconomyworkspace.slack.com/archives/C05E8EYSJ3E/p1703605643390259 when we provide incorrect gas values during sendUserOp our bundler is currently (incorrectly) setup in a way which loses money, so we should error if we can't find them. I would remove the try catch here because of this. Please someone cmiiw

Reply:
yes we can just throw. in buildUserOp we can catch this and rethrow. If we want to handle differently later with strictMode it can be easily done imo